frontend/leaderboard/src/Sheets.re

```diff
@@ -2,54 +2,92 @@ open SheetsBinding;
+  | exception (Js.Exn.Error(_)) => None
+  };
+};
 
-let authUrlConfig = {access_type: "offline", scope: scopes};
+let requestURICode = (client, cb) => {
+  let authUrlConfig = {access_type: "offline", scope: scopes};
+  let authUrl = generateAuthUrl(~client, ~authUrlConfig);
+  Js.log("Authorize this app by visiting this url: " ++ authUrl);
 
-let sheetsQuery = {
-  spreadsheetId: "1BxiMVs0XRA5nFMdKvBdBZjgmUUqptlbs74OgvE2upms",
-  range: "Class Data!A2:E",
-};
+  let readLine =
+    createInterface(interfaceOptions(~input=[%raw "process.stdin"]));
 
-let getSheets = token => {
-  setCredentials(~client, ~token);
-  get(
-    ~sheets,
-    ~sheetsQuery,
-    (~error, ~res) => {
-      Js.log(error);
-      Js.log(res.data.values);
+  question(
+    readLine,
+    "Enter the code from that page here: ",
+    code => {
+      switch (code) {
+      | "" => cb(Error(code))
+      | _ => cb(Ok(code))
+      };
+      close(readLine);
     },
   );
 };
 
-let createToken = () => {
-  getToken(
-    ~client,
-    ~code,
-    (~error, ~token) => {
-      Js.log(error);
+let createToken = (client, code, cb) => {
+  getToken(~client, ~code, (~error, ~token) => {
+    switch (Js.Nullable.toOption(error)) {
+    | Some(error) => cb(Error(error))
+    | None =>
       setCredentials(~client, ~token);
       Js.Json.stringify(token) |> Node.Fs.writeFileAsUtf8Sync(tokenPath);
-    },
-  );
+      Js.log("Token stored too: " ++ tokenPath);
+      cb(Ok());
+    }
+  });
 };
 
-let printSheets = () => {
-  switch (Node.Fs.readFileAsUtf8Sync(tokenPath)) {
-  | token => token |> Js.Json.parseExn |> getSheets
-  | exception (Js.Exn.Error(_)) => createToken()
-  };
+let getRange = (client, sheetsQuery, cb) => {
+  let sheets = sheets({version: "v4", auth: client});
+
+  get(~sheets, ~sheetsQuery, (~error, ~res) => {
+    switch (Js.Nullable.toOption(error)) {
+    | None => cb(Ok(res.data.values))
+    | Some(error) =>
+      Js.log("get() returned an error: " ++ error);
+      cb(Error(error));
+    }
+  });
 };
 
-// uncomment to get a valid code
-//Js.log(generateAuthUrl(~client, ~authUrlConfig));
+let createClient = (clientCredentials, cb) => {
+  let {clientId, clientSecret, redirectURI} = clientCredentials;
+  let client = oAuth2(~clientId, ~clientSecret, ~redirectURI);
 
-printSheets();
\ No newline at end of file
+  switch (readAndParseToken()) {
+  | Some(token) =>
+    setCredentials(~client, ~token);
+    cb(Ok(client));
+  | None =>
+    Js.log("Error loading client token file");
+    requestURICode(
+      client,
+      fun
+      | Ok(code) => {
+          createToken(
+            client,
+            code,
+            fun
+            | Ok () => cb(Ok(client))
+            | Error(error) => {
+                Js.log(
+                  "Error while trying to retrieve access token: " ++ error,
+                );
+                cb(Error(client));
+              },
+          );
+        }
+      | Error(_) => cb(Error(client)),
+    );
```

Comment:
Is the use of _fun_ preferred here or would there be a better way of passing in functions from a syntactic sense?

Reply:
fun is ok, `(c) => switch (c) { ... }` is equivalent and also ok. `fun` has slightly weird syntax as you may have discovered.

Reply:
Yeah, the syntax is a little different then what I'm used too. I think, for this reason, I'll switch to the other way you described.